implants/golem/src/inter/eval.rs

```diff
@@ -0,0 +1,304 @@
+        let mut builtin_symbols: HashMap<String, LspUrl> = HashMap::new();
+        for doc in get_registered_docs() {
+            let uri = Self::url_for_doc(&doc);
+            builtin_symbols.insert(doc.id.name.clone(), uri.clone());
+            builtins.entry(uri).or_default().push(doc);
+        }
+        let builtin_docs = builtins
+            .into_iter()
+            .map(|(u, ds)| (u, render_docs_as_code(&ds).join("\n\n")))
+            .collect();
+
+        Ok(Self {
+            mode,
+            print_non_none,
+            prelude,
+            module,
+            builtin_docs,
+            builtin_symbols,
+        })
+    }
+
+    fn url_for_doc(doc: &Doc) -> LspUrl {
+        let url = match &doc.item {
+            DocItem::Module(_) => Url::parse("starlark:/native/builtins.bzl").unwrap(),
+            DocItem::Object(_) => {
+                Url::parse(&format!("starlark:/native/builtins/{}.bzl", doc.id.name)).unwrap()
+            }
+            DocItem::Function(_) => Url::parse("starlark:/native/builtins.bzl").unwrap(),
+        };
+        LspUrl::try_from(url).unwrap()
+    }
+
+    fn new_module(prelude: &[FrozenModule]) -> Module {
+        let module = Module::new();
+        for p in prelude {
+            module.import_public_symbols(p);
+        }
+        module
+    }
+
+    fn go(&self, file: &str, ast: AstModule) -> EvalResult<impl Iterator<Item = EvalMessage>> {
+        let warnings: Either<std::iter::Empty<EvalMessage>, std::iter::Empty<EvalMessage>> = Either::Left(iter::empty());
+        let mut errors = Either::Left(iter::empty());
+        if errors.is_left() { } // Supress errors warning. We're no longer using a check mode so this code is dead.
+        let final_ast = match self.mode {
+            ContextMode::Run => {
+                errors = Either::Right(self.run(file, ast).messages);
+                None
+            }
+        };
+        EvalResult {
+            messages: warnings.chain(errors),
+            ast: final_ast,
+        }
+    }
+
+    // Convert an anyhow over iterator of EvalMessage, into an iterator of EvalMessage
+    fn err(
+        file: &str,
+        result: anyhow::Result<EvalResult<impl Iterator<Item = EvalMessage>>>,
+    ) -> EvalResult<impl Iterator<Item = EvalMessage>> {
+        match result {
+            Err(e) => EvalResult {
+                messages: Either::Left(iter::once(EvalMessage::from_anyhow(Path::new(file), &e))),
+                ast: None,
+            },
+            Ok(res) => EvalResult {
+                messages: Either::Right(res.messages),
+                ast: res.ast,
+            },
+        }
+    }
+
+    pub(crate) fn expression(
+        &self,
+        content: String,
+    ) -> EvalResult<impl Iterator<Item = EvalMessage>> {
+        let file = "expression";
+        Self::err(
+            file,
+            AstModule::parse(file, content, &dialect()).map(|module| self.go(file, module)),
```

Comment:
Seems writing to stderr happens in the starlark library.
Going to move on for now. Will look into the starlark code base and maybe open an issue there.